src/zkevm_specs/evm/execution/memory_copy.py

```diff
@@ -0,0 +1,53 @@
+from ...util import FQ, N_BYTES_MEMORY_SIZE
+from ..execution_state import ExecutionState
+from ..instruction import Instruction, Transition
+from ..step import CopyToMemoryAuxData
+from ..table import RW, TxContextFieldTag
+from ..util import BufferReaderGadget
+
+
+MAX_COPY_BYTES = 71
+
+
+def copy_to_memory(instruction: Instruction):
+    aux = instruction.curr.aux_data
+    assert isinstance(aux, CopyToMemoryAuxData)
+
+    buffer_reader = BufferReaderGadget(
+        instruction, MAX_COPY_BYTES, aux.src_addr, aux.src_addr_end, aux.bytes_left
+    )
+
+    data = []
+    rw_counter_delta = 0
+    for i in range(MAX_COPY_BYTES):
+        if not buffer_reader.read_flag(i):
+            byte = FQ.zero()
+        elif aux.from_tx == 1:
+            byte = instruction.tx_context_lookup(
+                aux.tx_id, TxContextFieldTag.CallData, aux.src_addr + i
+            )
+        else:
+            byte = instruction.memory_lookup(RW.Read, aux.src_addr + i)
+        buffer_reader.constrain_byte(i, byte)
+        if buffer_reader.has_data(i):
+            assert byte == instruction.memory_lookup(RW.Write, aux.dst_addr + i)
```

Comment:
Shouldn't this be a `constraint_equal`?

Reply:
Updated

Reply:
Just would like a double check with @han0110 in case I'm wrong on my assumption.

Reply:
yeah, it's convenient to throw the same error by calling `constrain_equal`, but currently that doens't used actually (failure case tests)

Reply:
The `constraint_equal` I meant was at circuit level rather than a python assertion.

That was my concern, if at circuit level we should constraint both Cells equality. Whether we needed to add this or not :smile:

Reply:
Ah I see, in circuit the `byte` will be used as lookup's input, so it won't explicitly call the `constraint_equal` actually.

Reply:
Yeah, that's because in spec we don't actually have "witness assignment", so we have no way to come up with the `byte`, then it becomes the output of the lookup.

Reply:
Ohh that's right!! Perfect then! No issues from my side.